src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +748,103 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientManager.GetAvailableClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current).ConfigureAwait(false);
+
+                    _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!TryFailOver(ae, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    if (!TryFailOver(e, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private bool TryFailOver(Exception e, IEnumerator<ConfigurationClient> clientEnumerator)
+        {
+            if (!ShouldFailoverForException(e))
+            {
+                return false;
+            }
+
+            _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+            if (!clientEnumerator.MoveNext())
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private void UpdateCacheExpirationForWatchers(bool success)
+        {
+            // Update the cache expiration time for all refresh registered settings and feature flags
+            foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers.Concat(_options.MultiKeyWatchers))
+            {
+                UpdateCacheExpirationTime(changeWatcher, success);
+            }
+
+        }
+
+        private bool ShouldFailoverForException(Exception ex)
+        {
+            if (ex is AggregateException aggregateException)
+            {
+                IReadOnlyCollection<Exception> innerExceptions = aggregateException.InnerExceptions;
+
+                if (innerExceptions != null && innerExceptions.Any() && innerExceptions.All(ex => ex is RequestFailedException))
+                {
+                    return ShouldFailOver(innerExceptions.Last() as RequestFailedException);
+                }
+            }
+            else if (ex is RequestFailedException rfe)
+            {
+                return ShouldFailOver(rfe);
+            }
+
+            return false;
+        }
+
+        private bool ShouldFailOver(RequestFailedException rfe)

```

Comment:
Maybe a better name for this function is `IsFailOverable` so it doesn't indicate whether we should or can fail over. It simply indicates the errors are eligible for failing over.